public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts

```diff
@@ -0,0 +1,526 @@
+type NodeTypeName =
+  | '⚠' // this is used as error-name
+  | 'Expr'
+  | 'Grouping'
+  | 'Identifier'
+  | 'LogExpr'
+  | 'LogRangeExpr'
+  | 'LogQL'
+  | 'Matcher'
+  | 'Matchers'
+  | 'PipelineStage'
+  | 'PipelineExpr'
+  | 'Range'
+  | 'RangeAggregationExpr'
+  | 'Selector'
+  | 'String'
+  | 'VectorAggregationExpr';
+
+type Path = Array<[Direction, NodeTypeName]>;
+
+function move(node: SyntaxNode, direction: Direction): SyntaxNode | null {
+  return node[direction];
+}
+
+function walk(node: SyntaxNode, path: Path): SyntaxNode | null {
+  let current: SyntaxNode | null = node;
+  for (const [direction, expectedType] of path) {
+    current = move(current, direction);
+    if (current === null) {
+      // we could not move in the direction, we stop
+      return null;
+    }
+    if (current.type.name !== expectedType) {
+      // the reached node has wrong type, we stop
+      return null;
+    }
+  }
+  return current;
+}
+
+function getNodeText(node: SyntaxNode, text: string): string {
+  return text.slice(node.from, node.to);
+}
+
+function parseStringLiteral(text: string): string {
+  // If it is a string-literal, it is inside quotes of some kind
+  const inside = text.slice(1, text.length - 1);
+
+  // Very simple un-escaping:
+
+  // Double quotes
+  if (text.startsWith('"') && text.endsWith('"')) {
+    // NOTE: this is not 100% perfect, we only unescape the double-quote,
+    // there might be other characters too
+    return inside.replace(/\\"/, '"');
+  }
+
+  // Single quotes
+  if (text.startsWith("'") && text.endsWith("'")) {
+    // NOTE: this is not 100% perfect, we only unescape the single-quote,
+    // there might be other characters too
+    return inside.replace(/\\'/, "'");
+  }
+
+  // Backticks
+  if (text.startsWith('`') && text.endsWith('`')) {
+    return inside;
+  }
+
+  throw new Error(`Invalid string literal: ${text}`);
+}
+
+export type LabelOperator = '=' | '!=' | '=~' | '!~';
+
+export type Label = {
+  name: string;
+  value: string;
+  op: LabelOperator;
+};
+
+export type Situation =
+  | {
+      type: 'EMPTY';
+    }
+  | {
+      type: 'AT_ROOT';
+    }
+  | {
+      type: 'IN_DURATION';
+    }
+  | {
+      type: 'IN_AGGREGATION';
+    }
+  | {
+      type: 'IN_GROUPING';
+      otherLabels: Label[];
+    }
+  | {
+      type: 'IN_LABEL_SELECTOR_NO_LABEL_NAME';
+      otherLabels: Label[];
+    }
+  | {
+      type: 'IN_LABEL_SELECTOR_WITH_LABEL_NAME';
+      labelName: string;
+      betweenQuotes: boolean;
+      otherLabels: Label[];
+    }
+  | {
+      type: 'AFTER_SELECTOR';
+      afterPipe: boolean;
+      labels: Label[];
+    };
+
+type Resolver = {
+  path: NodeTypeName[];
+  fun: (node: SyntaxNode, text: string, pos: number) => Situation | null;
+};
+
+function isPathMatch(resolverPath: string[], cursorPath: string[]): boolean {
+  return resolverPath.every((item, index) => item === cursorPath[index]);
+}
+
+const ERROR_NODE_NAME: NodeTypeName = '⚠'; // this is used as error-name
```

Comment:
```suggestion
const ERROR_NODE_NAME: NodeTypeId = 0; // this is used as error-id
```

Reply:
Good catch. Refactored everything to use node identifiers.